internal/cmdrpkgupdate/discover.go

```diff
@@ -130,18 +147,81 @@ func (r *runner) getRepositories() (*configapi.RepositoryList, error) {
 }
 
 // fetches all package revision numbers for packages with the name upstreamPackageName from the repo
-func (r *runner) getUpstreamRevisions(repo configapi.Repository, upstreamPackageName string) []string {
-	var result []string
+func (r *runner) getUpstreamRevisions(repo configapi.Repository, upstreamPackageName string) []porchapi.PackageRevision {
+	var result []porchapi.PackageRevision
 	for _, pkgRev := range r.prs {
 		if pkgRev.Spec.Lifecycle != porchapi.PackageRevisionLifecyclePublished {
 			// only consider published packages
 			continue
 		}
-		if pkgRev.Spec.RepositoryName != repo.Name ||
-			pkgRev.Spec.PackageName != upstreamPackageName {
-			continue
+		if pkgRev.Spec.RepositoryName == repo.Name && pkgRev.Spec.PackageName == upstreamPackageName {
+			result = append(result, pkgRev)
 		}
-		result = append(result, pkgRev.Spec.Revision)
 	}
 	return result
 }
+
+func printupstreamUpdates(upstreamUpdates [][]string, cmd *cobra.Command) error {
+	w := printers.GetNewTabWriter(cmd.OutOrStdout())
+	if _, err := fmt.Fprintln(w, "PACKAGE REVISION\tUPSTREAM REPOSITORY\tUPSTREAM UPDATES"); err != nil {
+		return err
+	}
+	for _, pkgRev := range upstreamUpdates {
+		if _, err := fmt.Fprintln(w, strings.Join(pkgRev, "\t")); err != nil {
+			return err
+		}
+	}
+	return w.Flush()
+}
+
+func printdownstreamUpdates(downstreamUpdatesMap map[string][]porchapi.PackageRevision, args []string, cmd *cobra.Command) error {
+	var downstreamUpdates [][]string
+	for upstreamPkgRev, downstreamPkgRevs := range downstreamUpdatesMap {
+		split := strings.Split(upstreamPkgRev, ":")
+		upstreamPkgRevName := split[0]
+		upstreamPkgRevNum := split[1]
+		for _, downstreamPkgRev := range downstreamPkgRevs {
+			// figure out which upstream revision the downstream revision is based on
+			lastIndex := strings.LastIndex(downstreamPkgRev.Status.UpstreamLock.Git.Ref, "v")
+			if lastIndex < 0 {
+				// this ref isn't formatted the way that porch expects
+				continue
+			}
+			downstreamRev := downstreamPkgRev.Status.UpstreamLock.Git.Ref[lastIndex:]
+			downstreamUpdates = append(downstreamUpdates, []string{upstreamPkgRevName, upstreamPkgRevNum, downstreamPkgRev.Name, downstreamRev})
+		}
+	}
+
+	w := printers.GetNewTabWriter(cmd.OutOrStdout())
+	var pkgRevsToPrint [][]string
+	if len(args) != 0 {
+		for _, arg := range args {
+			for _, pkgRev := range downstreamUpdates {
+				// filter out irrelevant packages based on provided args
+				if arg != pkgRev[0] {
+					continue
+				}
+				pkgRevsToPrint = append(pkgRevsToPrint, pkgRev)
+			}
+		}
+	} else {
+		pkgRevsToPrint = downstreamUpdates
+	}
+
+	if len(pkgRevsToPrint) == 0 {
+		if _, err := fmt.Fprintln(w, "All downstream packages are up to date."); err != nil {
+			return err
+		}
+	} else {
+		if _, err := fmt.Fprintln(w, "UPSTREAM PACKAGE\tUPSTREAM REVISION\tDOWNSTREAM PACKAGE\tDOWNSTREAM REVISION"); err != nil {
```

Comment:
The last column "DOWNSTREAM REVISION" isn't actually the package revision number of the downstream revision, it's the upstream revision number that the downstream package revision is based on - but I was struggling to think of a clear, concise way to specify that in the column name. It'll probably be good to document it somewhere what these column names actually mean.

Reply:
Yeah, I agree the terminology is a bit difficult here. 

When the user specifies `--discovery=upstream`, we currently list packages (using just "PACKAGE" as the column heading). I think maybe we should just use the heading "PACKAGE" in the `--discovery=downstream` case as well, since having columns named both "UPSTREAM PACKAGE" and "DOWNSTREAM PACKAGE" becomes a bit confusing.

For the "DOWNSTREAM REVISION" column, could we maybe name it something like "FROM VERSION" or "INCLUDES VERSION" to make it clearer that this is the version of the upstream that is included in the downstream package.

Reply:
> When the user specifies --discovery=upstream, we currently list packages (using just "PACKAGE" as the column heading). I think maybe we should just use the heading "PACKAGE" in the --discovery=downstream case as well, since having columns named both "UPSTREAM PACKAGE" and "DOWNSTREAM PACKAGE" becomes a bit confusing.

👍 I will make it consistent with `--upstream`

> For the "DOWNSTREAM REVISION" column, could we maybe name it something like "FROM VERSION" or "INCLUDES VERSION" to make it clearer that this is the version of the upstream that is included in the downstream package.

I like "FROM VERSION", but taking it a bit further, maybe we could combine the revision columns:

```
$ kpt alpha rpkg update --discover=downstream
PACKAGE REVISION                                    DOWNSTREAM PACKAGE                                          AVAILABLE UPDATE
kpt-samples-b2d61a19750d212c61a1c64b0a1cf1ce0efddc2 test-deployments-2f36052b298673c231e8debcdd1a4553200b45c7   v1->v2
```

WDYT?

Reply:
yeah, I like that idea.